Lib/multiprocessing/shared_memory.py

```diff
@@ -1,225 +1,250 @@
+"""Provides shared memory for direct access across processes.
+
+The API of this package is currently provisional. Refer to the
+documentation for details.
+"""
 
 
 __all__ = [ 'SharedMemory', 'PosixSharedMemory', 'WindowsNamedSharedMemory',
-            'ShareableList', 'shareable_wrap',
-            'SharedMemoryServer', 'SharedMemoryManager', 'SharedMemoryTracker' ]
+            'ShareableList',
+            'SharedMemoryServer', 'SharedMemoryManager' ]
 
 
-from functools import reduce
+from functools import partial
 import mmap
-from .managers import DictProxy, SyncManager, Server
+from .managers import dispatch, BaseManager, Server, State, ProcessError
 from . import util
 import os
-import random
 import struct
-import sys
+import secrets
 try:
-    from _posixshmem import _PosixSharedMemory, Error, ExistentialError, O_CREX
+    import  _posixshmem
+    from os import O_CREAT, O_EXCL, O_TRUNC
 except ImportError as ie:
-    if os.name != "nt":
-        # On Windows, posixshmem is not required to be available.
-        raise ie
-    else:
-        _PosixSharedMemory = object
-        class ExistentialError(BaseException): pass
-        class Error(BaseException): pass
-        O_CREX = -1
+    O_CREAT, O_EXCL, O_TRUNC = 64, 128, 512
+
+O_CREX = O_CREAT | O_EXCL
+
+if os.name == "nt":
+    import _winapi
+
+    PAGE_READONLY = 0x02
+    PAGE_EXECUTE_READWRITE = 0x04
+    INVALID_HANDLE_VALUE = -1
 
 
 class WindowsNamedSharedMemory:
 
-    def __init__(self, name, flags=None, mode=None, size=None, read_only=False):
+    def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name is None:
-            name = f'wnsm_{os.getpid()}_{random.randrange(100000)}'
+            name = _make_filename()
+
+        if size == 0:
+            # Attempt to dynamically determine the existing named shared
+            # memory block's size which is likely a multiple of mmap.PAGESIZE.
+            try:
+                h_map = _winapi.OpenFileMappingW(PAGE_READONLY, False, name)
+            except OSError:
+                raise FileNotFoundError(name)
+            try:
+                p_buf = _winapi.MapViewOfFile(h_map, PAGE_READONLY, 0, 0, 0)
+            finally:
+                _winapi.CloseHandle(h_map)
+            size = _winapi.VirtualQuerySize(p_buf)
+
+        if flags == O_CREX:
+            # Create and reserve shared memory block with this name until
+            # it can be attached to by mmap.
+            h_map = _winapi.CreateFileMappingW(
+                INVALID_HANDLE_VALUE,
+                _winapi.NULL,
+                PAGE_EXECUTE_READWRITE,
+                (size >> 32) & 0xFFFFFFFF,
+                size & 0xFFFFFFFF,
+                name
+            )
+            try:
+                last_error_code = _winapi.GetLastError()
+                if last_error_code == _winapi.ERROR_ALREADY_EXISTS:
+                    raise FileExistsError(f"File exists: {name!r}")
```

Comment:
It would be nice to include the `errno` and `winerror` values for completeness: `FileExistsError(errno.EEXIST, os.strerror(errno.EEXIST), name, _winapi.ERROR_ALREADY_EXISTS)`.

Reply:
Good point -- added in commit 68785333a25d902a42252ace3a8268357efd6094.